## kraken2-nf 
a Nextflow pipeline for running [kraken2](https://github.com/DerrickWood/kraken2) classification of metagenomes - from building kraken2-database to classification.

## How the pipeline works in details
The main workflow can be found in [main.nf](https://github.com/Phuong-Le/bowtie2-nf/blob/main/main.nf)

1. If any of the 3 kraken2-databse files (.k2d) generated by `kraken2-build --build` do not exist, then build kraken2 database. Otherwise, proceed to classification. Source code is in [modules/buildDB.nf](https://github.com/Phuong-Le/kraken2-nf/blob/main/modules/buildDB.nf) 

2. Classification of metagenome based on the fastq files and the kraken2 database generated above. Source code is in [modules/classify.nf](https://github.com/Phuong-Le/kraken2-nf/blob/main/modules/classify.nf) 


## Dependencies
- [Nextflow](https://www.nextflow.io/)
- [Docker](https://www.docker.com/) if using own machine or [Singularity](https://sylabs.io/singularity/) if using a shared HPC
- [`kraken2`](https://github.com/DerrickWood/kraken2)


## Installation
- Clone this repo
``` 
git clone git@github.com:Phuong-Le/kraken2-nf.git
```


## Usage
```
nf_script=/path/to/main.nf
config_file=/path/to/customed_config/file #eg lsf.config
db=/path/to/database/dir
sample_params=/path/to/sample/params/file # a tsv file with 3 columns, sample name, path to fastq1 and path to fastq2
outdir=/path/to/outdir # after execution, outdir should contain the kraken reports for all samples 

nextflow run ${nf_script} -c ${config_file} \
    --db ${db} --sample_params ${sample_params} --outdir ${outdir}
```

example on an lsf system like at Sanger
```
bsub -cwd /path/to/working_dir -o %J.out -e %J.err -R "select[mem>1000] rusage[mem=1000]" -M1000 \
    "nextflow run ${nf_script} -c ${config_file} \
    --db ${db} --sample_params ${sample_params} --outdir ${outdir}"
```

demo file for `${sample_params}` is found in [demo_files](https://github.com/Phuong-Le/kraken2-nf/blob/main/demo_files/sample_params.tsv)


## Authors 
[Phuong Le](https://github.com/Phuong-Le) (email: al35@sanger.ac.uk) 

## Acknowledgements
[Vicky Carr](https://github.com/blue-moon22)

## Scope for the future
The pipeline hasn't fully addressed building kraken2 database - users still need to manually include references and download their taxonomy. Future development could try to design the best way to automate this process. In the meantime, following [kraken2 manual](https://github.com/DerrickWood/kraken2/wiki/Manual#custom-databases), users can either download the standard databse
```
kraken2-build --download-library ${standard_reference} --db $DBNAME 
```
`${standard_reference}` is one of the following: archaea, bacteria, plasmid, viral, human, fungi, plant, protozoa, nr, nt, UniVec, UniVec_Core, env_nr, env_nt

or add customed reference, provided the fasta sequence ID is an NCBI accession number
```
kraken2-build --add-to-library chr1.fa --db $DBNAME
```

Once the above step is finished, there will be a subfolder called "library" under `${DBNAME}` directory. Do the following to download the taxonomy
```
kraken2-build --download-taxonomy --db $DBNAME
```

${DBNAME} is now ready for our `kraken2-nf` pipeline